File: src/main/scala/com/kazzla/domain/irpc/Transferable.scala
/* Copyright (C) 2012 BJöRFUAN
 * This source and related resources are distributed under Apache License, Version 2.0.
 */
package com.kazzla.domain.irpc

import java.nio.ByteBuffer
import com.kazzla.debug._
import scala.Some

// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Transferable
// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
sealed abstract class Transferable private[irpc]() extends java.io.Serializable

// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Open
// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
final case class Open(id:Long, timeout:Long, callback:Boolean, name:String, args:Any*) extends Transferable {

	// ========================================================================
	// インスタンスの文字列化
	// ========================================================================
	/**
	 * このインスタンスを文字列化します。
	 * @return インスタンスの文字列
	 */
	override def toString():String = {
		"Open[" + id + "](" + timeout + "," + callback + "," + name + '(' + makeDebugString(args) + ')'
	}

}

// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Close
// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
final case class Close(id:Long, code:Close.Code.Value, message:String, args:Any*) extends Transferable {

	// ========================================================================
	// インスタンスの文字列化
	// ========================================================================
	/**
	 * このインスタンスを文字列化します。
	 * @return インスタンスの文字列
	 */
	override def toString():String = {
		"Close[" + id + "](" + (code match {
			case Close.Code.CLOSE => com.kazzla.debug.makeDebugString(result)
			case _ => message
		}) + ")"
	}

}
object Close {
	object Code extends Enumeration {
		val CLOSE = Value(1)
		val CANCEL = Value(2)
		val ERROR = Value(3)
		val FATAL = Value(127)
	}
}

// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Block
// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
final case class Block(id:Long, sequence:Int, binary:Array[Byte]) extends Transferable {

	// ========================================================================
	// インスタンスの文字列化
	// ========================================================================
	/**
	 * このインスタンスを文字列化します。
	 * @return インスタンスの文字列
	 */
	override def toString():String = {
		id + ":[" + sequence + ":" + makeDebugString(binary, 25) + "]"
	}

}

// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Control
// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
final case class Control(id:Long, code:Byte, args:Any*) extends Transferable {

	// ========================================================================
	// インスタンスの文字列化
	// ========================================================================
	/**
	 * このインスタンスを文字列化します。
	 * @return インスタンスの文字列
	 */
	override def toString():String = {
		id + ":{" + "%05d".format(code) + ":" + makeDebugString(args) + "}"
	}

}

File: src/main/java/com/kazzla/domain/irpc/Transferable.java
/* Copyright (C) 2012 BJöRFUAN
 * This source and related resources are distributed under Apache License, Version 2.0.
 */
package com.kazzla.domain.irpc;

import sun.rmi.transport.ObjectTable;

import java.util.Map;

// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// Transferable
// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
public abstract class Transferable implements java.io.Serializable{
	public final long pipeId;
	Transferable(long pipeId){
		this.pipeId = pipeId;
	}

	// ========================================================================
	// パラメータの確認
	// ========================================================================
	/**
	 * 指定されたパラメータが iRPC 仕様に沿ったものかを確認します。
	 * このメソッドは未定義の型のオブジェクトを検出した場合に例外を発生させますが、プロダク
	 * ション環境では実行されないないように assert と組み合わせて使用することを想定してる
	 * ため boolean を返します。
	 */
	@SuppressWarnings("unchecked")
	public static boolean verifyArgumentTypes(Object... args){
		for(Object arg: args){
			if(arg == null
				|| arg instanceof Boolean || arg instanceof Byte || arg instanceof Short
				|| arg instanceof Integer || arg instanceof Long || arg instanceof Float
				|| arg instanceof Double || arg instanceof String){
				// do nothing
			} else if(arg instanceof Map){
				verifyArgumentTypes(((Map<Object,Object>)arg).keySet());
				verifyArgumentTypes(((Map<Object,Object>)arg).values());
			} else if(arg instanceof Iterable){
				for(Object elem: (Iterable<Object>)arg){
					verifyArgumentTypes(elem);
				}
			} else if(arg instanceof Object[]){
				for(Object elem: (Object[])arg){
					verifyArgumentTypes(elem);
				}
			} else {
				throw new IllegalArgumentException("unsupported argument type: " + arg.getClass().getName());
			}
		}
		return true;
	}

}